community/kernel/src/main/java/org/neo4j/kernel/configuration/Config.java

```diff
@@ -74,237 +73,467 @@
 
         validators.addAll( additionalValidators );
         migrator = new AnnotationBasedConfigurationMigrator( settingsClasses );
+        this.overriddenDefaults.putAll( overriddenDefaults );
 
-        Map<String,String> settings = initSettings( configFile, settingsPostProcessor, overriddenSettings, this.log );
-        replaceSettings( settings, configFile.isPresent() );
+        boolean fromFile = configFile != null;
+        if ( fromFile )
+        {
+            loadFromFile( configFile, log ).forEach( initialSettings::putIfAbsent );
+        }
+
+        overriddenDefaults.forEach( initialSettings::putIfAbsent );
+
+        migrateAndValidateAndUpdateSettings( initialSettings, fromFile );
+
+        // Only warn for deprecations if red from a file
+        if ( fromFile )
+        {
+            warnAboutDeprecations( params );
+        }
+    }
+
+    /**
+     * Retrieves a configuration value. If no value is configured, a default value will be returned instead. Note that
+     * {@code null} is a valid value.
+     *
+     * @param setting The configuration property.
+     * @param <T> the underlying type of the setting.
+     * @return the value of the given setting, {@code null} can be returned.
+     */
+    @Override
+    public <T> T get( Setting<T> setting )
+    {
+        return setting.apply( params::get );
     }
 
     /**
-     * Returns a copy of this config with the given modifications.
+     * Test whether a setting is configured or not. Can be used to check if default value will be returned or not.
      *
-     * @return a new modified config, leaves this config unchanged.
+     * @param setting The setting to check.
+     * @return {@code true} if the setting is configures, {@code false} otherwise implying that the default value will
+     * be returned if applicable.
      */
-    public Config with( Map<String,String> additionalConfig )
+    public boolean isConfigured( Setting<?> setting )
     {
-        Map<String,String> newParams = new HashMap<>( params ); // copy is returned
-        newParams.putAll( additionalConfig );
-        return new Config( Optional.empty(), newParams, NO_POST_PROCESSING, validators, Optional.of( log ) );
+        return params.containsKey( setting.name() );
     }
 
     /**
-     * Returns a copy of this config with the given modifications except for any settings which already have a
-     * specified value.
+     * Extract all identifiers within a group, e.g. giving that prefix is {@code "dbms.ssl.policy"}), the
+     * setting {@code "dbms.ssl.policy.default.base_directory"} will return {@code "default"}.
      *
-     * @return a new modified config, leaves this config unchanged.
+     * @param prefix group prefix for the settings interested in
+     * @return A set of uniquely and sorted group keys.
      */
-    public Config withDefaults( Map<String,String> additionalDefaults )
+    public Set<String> identifiersFromPrefix( String prefix )
```

Comment:
We use the match to extract the identifier. Javadoc is updated to make this more clear.